pkg/storage/tsdb/bucketindex/writer.go

```diff
@@ -0,0 +1,273 @@
+		return nil, errors.Wrap(err, "gzip bucket index")
+	}
+	if err := gzip.Close(); err != nil {
+		return nil, errors.Wrap(err, "close gzip bucket index")
+	}
+
+	// Upload the index to the storage.
+	if err := w.bkt.Upload(ctx, IndexCompressedFilename, &gzipContent); err != nil {
+		return nil, errors.Wrap(err, "upload bucket index")
+	}
+
+	return idx, nil
+}
+
+// GenerateIndex generates the bucket index and returns it, without storing it to the storage.
+// If the old index is not passed in input, then the bucket index will be generated from scratch.
+func (w *Writer) GenerateIndex(ctx context.Context, old *Index) (*Index, error) {
+	var oldBlocks []*Block
+	var oldBlockDeletionMarks []*BlockDeletionMark
+
+	// Read the old index, if provided.
+	if old != nil {
+		oldBlocks = old.Blocks
+		oldBlockDeletionMarks = old.BlockDeletionMarks
+	}
+
+	blocks, err := w.generateBlocksIndex(ctx, oldBlocks)
+	if err != nil {
+		return nil, err
+	}
+
+	blockDeletionMarks, err := w.generateBlockDeletionMarksIndex(ctx, oldBlockDeletionMarks)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Index{
+		Version:            IndexVersion1,
+		Blocks:             blocks,
+		BlockDeletionMarks: blockDeletionMarks,
+		UpdatedAt:          time.Now().Unix(),
+	}, nil
+}
+
+func (w *Writer) generateBlocksIndex(ctx context.Context, old []*Block) ([]*Block, error) {
+	out := make([]*Block, 0, len(old))
+	discovered := map[ulid.ULID]struct{}{}
+
+	// Find all blocks in the storage.
+	err := w.bkt.Iter(ctx, "", func(name string) error {
+		if id, ok := block.IsBlockDir(name); ok {
+			discovered[id] = struct{}{}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "list blocks")
+	}
+
+	// Since blocks are immutable, all blocks already existing in the index can just be copied.
+	for _, b := range old {
+		if _, ok := discovered[b.ID]; ok {
+			out = append(out, b)
+			delete(discovered, b.ID)
+		}
+	}
+
+	// Remaining blocks are new ones and we have to fetch the meta.json for each of them, in order
+	// to find out if their upload has been completed (meta.json is uploaded last) and get the block
+	// information to store in the bucket index.
+	for id := range discovered {
+		b, err := w.generateBlockIndexEntry(ctx, id)
+		if errors.Is(err, ErrBlockMetaNotFound) {
+			level.Warn(w.logger).Log("msg", "skipped partial block when generating bucket index", "block", id.String())
+			continue
+		}
+		if errors.Is(err, ErrBlockMetaCorrupted) {
+			level.Error(w.logger).Log("msg", "skipped block with corrupted meta.json when generating bucket index", "block", id.String(), "err", err)
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		out = append(out, b)
+	}
+
+	return out, nil
+}
+
+func (w *Writer) generateBlockIndexEntry(ctx context.Context, id ulid.ULID) (*Block, error) {
+	metaFile := path.Join(id.String(), block.MetaFilename)
+
+	// Get the block's meta.json file.
+	r, err := w.bkt.Get(ctx, metaFile)
```

Comment:
comment: I'm bit surprised to see that we don't use some Thanos function for reading meta.json file. But I cannot find such public function in Thanos. :confused:

Reply:
There's `metadata.Read()` and `metadata.ReadMarker()` from Thanos but:
- with `metadata.Read()` we would have no way to distinguish a corrupted meta from a network error
- `metadata.ReadMarker()` requires an `InstrumentedBucket` and it's quite annoying to propagate till here (I have tried). I can try again, but I would prefer to keep it in a separate PR.

Reply:
We could check if the `error` returned by `metadata.Read()` is a `json.SyntaxError` but unfortunately that doesn't work because the returned error is a `MultiErr`. We would need this https://github.com/prometheus/prometheus/pull/8133 in.